pkg/aspect/query/shared/query.go

```diff
@@ -0,0 +1,220 @@
+}
+
+type PromptRunner interface {
+	Run() (string, error)
+}
+
+func Prompt(label string) PromptRunner {
+	return &promptui.Prompt{
+		Label: label,
+	}
+}
+
+type ConfirmationRunner interface {
+	Run() (string, error)
+}
+
+func Confirmation(question string) ConfirmationRunner {
+	return &promptui.Prompt{
+		Label:     question,
+		IsConfirm: true,
+	}
+}
+
+type SelectRunner interface {
+	Run() (int, string, error)
+}
+
+func Select(presetNames []string) SelectRunner {
+	return &promptui.Select{
+		Label: "Select a preset query",
+		Items: presetNames,
+	}
+}
+
+func GetPrecannedQueries(verb string) []*PresetQuery {
+	// TODO: Queries should be loadable from the plugin config
+	// https://github.com/aspect-build/aspect-cli/issues/98
+	presets := []*PresetQuery{
+		{
+			Name:        "why",
+			Description: "Determine why targetA depends on targetB",
+			Query:       "somepath(?targetA, ?targetB)",
+			Verb:        "query",
+		},
+		{
+			Name:        "deps",
+			Description: "Get the deps of a target",
+			Query:       "deps(?target)",
+			Verb:        "query",
+		},
+		{
+			Name:        "adeps",
+			Description: "Get the deps of a target",
+			Query:       "deps(?target)",
+			Verb:        "aquery",
+		},
+		{
+			Name:        "cdeps",
+			Description: "Get the deps of a target",
+			Query:       "deps(?target)",
+			Verb:        "cquery",
+		},
+	}
+
+	switch verb {
+	case "query":
+		return filterPrecannedQueries("query", presets)
+	case "cquery":
+		return filterPrecannedQueries("cquery", presets)
+	case "aquery":
+		return filterPrecannedQueries("aquery", presets)
+	}
+
+	return presets
+}
+
+func filterPrecannedQueries(verb string, presets []*PresetQuery) []*PresetQuery {
+	filteredPresets := []*PresetQuery{}
+	for _, presetQuery := range presets {
+
+		if presetQuery.Verb == verb {
+			filteredPresets = append(filteredPresets, presetQuery)
+		}
+	}
+
+	return filteredPresets
+}
+
+func ProcessQueries(presets []*PresetQuery) (map[string]*PresetQuery, []string, error) {
+	processedPresets := make(map[string]*PresetQuery)
+	presetNames := make([]string, len(presets))
+	for i, p := range presets {
+		if _, exists := processedPresets[p.Name]; exists {
+			err := fmt.Errorf("duplicated preset query name %q", p.Name)
+			return processedPresets, presetNames, err
+		}
+		processedPresets[p.Name] = p
+		presetNames[i] = fmt.Sprintf("%s: %s", p.Name, p.Description)
+	}
+
+	return processedPresets, presetNames, nil
+}
+
+func RunQuery(bzl bazel.Bazel, verb string, query string) error {
+	bazelCmd := []string{
+		verb,
+		query,
+	}
+
+	if exitCode, err := bzl.Spawn(bazelCmd); exitCode != 0 {
+		err = &aspecterrors.ExitError{
+			Err:      err,
+			ExitCode: exitCode,
+		}
+		return fmt.Errorf("failed to run %q %q: %w", verb, query, err)
+	}
+
+	return nil
+}
+
+func ReplacePlaceholders(query string, args []string, p func(label string) PromptRunner) (string, error) {
+	placeholders := regexp.MustCompile(`(\?[a-zA-Z]*)`).FindAllString(query, -1)
```

Comment:
Move the `MustCompile` to the package scope since this step is expensive and the result of the compiled regex can be considered a "const".